### PR TITLE
Clinic: adds module to add patients

### PR DESCRIPTION
**Pull request by ashutoshgngwr**

### CODE REVIEW

File: clinic/src/clj/clinic/service/patient.clj [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+          resource (-> params
+                       (assoc :mrn mrn)
+                       (domain->fhir))
+          response (fc/create! fhir-server-url
+                               resource
+                               {"If-None-Exist" (str "identifier=" mrn-system "|" mrn)})
+          {status :status body :body} response]
+      (cond
+        (= 0 retry-count) (throw (ex-info "couldn't generate a unique MRN"
+                                          {:type :mrn-conflict}))
```
Comment (vkrmis) [inline clinic/src/clj/clinic/service/patient.clj]: You can _mostly_ get away with this in the real world, but when it breaks, the error here is difficult to handle cleanly and communicate to the user.

Some alternatives:
- Random uuids will [almost never collide.](https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions)
- You can generate a list of available ids and pop one off every time you need to assign one. You'll have to replenish this list from time to time so it doesn't run out

File: clinic/src/clj/clinic/routes/core.clj [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+(defroutes ^:private routes
+  (context "/api/v1" _
+    (context "/patients" _ patient/handler))
+  (not-found (r/status 404)))
+
+(defn- wrap-exception-handler [next-handler]
+  (fn exception-handler [request]
+    (try (next-handler request)
+         (catch Throwable t
+           (println "encountered unexpected error for request:" request)
```
Comment (vkrmis) [inline clinic/src/clj/clinic/routes/core.clj]: Might want to use a logger here, take a look at https://github.com/clojure/tools.logging

File: clinic/src/clj/clinic/fhir/utils.clj [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+(defn find-code
+  "Returns the value of the `code` attribute corresponding to the given `system`
+   in a FHIR Codeable Concept."
+  [system codeable-concept]
+  (some-> codeable-concept
+          (get :coding)
+          ((partial filter #(= system (get % :system))))
+          (first)
+          (get :code)))
```
Comment (vkrmis) [inline clinic/src/clj/clinic/fhir/utils.clj]: Nothing wrong with this ^, but you _could_ also write it this way which is more readable imo
```
(some->> codeable-concept
     :coding
     (filter #(= system (:system %))
     first
     :code)
```

File: clinic/src/clj/clinic/service/patient.clj [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    (when (params :marital-status)
+      (swap! resource assoc :maritalStatus {:coding [{:system marital-status-system
+                                                      :code (params :marital-status)}]}))
+    (when (params :email)
+      (swap! resource update :telecom conj {:system email-system
+                                            :value (params :email)}))
+    (when (params :phone)
+      (swap! resource update :telecom conj {:system phone-system
+                                            :value (params :phone)}))
+    @resource))
```
Comment (vkrmis) [inline clinic/src/clj/clinic/service/patient.clj]: Don't think the atom is needed here.
Can just thread the resource through a series of assocs:
```
(-> resource-map
    (cond-> (:marital-status params)
        (assoc :maritalStatus {:coding ...})
     ...
     )
``` 
There are cases when this isn't feasible and you actually want to maintain state in an atom, but you don't need it here. Using an atom like this is analogous to assigning a value to a variable and iteratively/procedurally mutating it.

Comment (ashutoshgngwr) [reply clinic/src/clj/clinic/service/patient.clj]: Didn't know about `cond->`. Refactored it!

File: clinic/src/clj/clinic/service/patient.clj (hunk 1 above)
Comment (ashutoshgngwr) [reply clinic/src/clj/clinic/service/patient.clj]: As we talked, I've removed the MRN field from the Patient entity.

**Comment by neenaoffline** [comment]
Let's merge this and do a post-merge review.

**Pull request merged**